### DISCUSSION TITLE
Mocks usage

### DISCUSSION THREAD

**Discussion by mamdouni** [Decathlon/tzatziki#17] [Show and tell]
Hello,

I would like to know more about how to mock an api call. 
In my case i have to test a service (called A for example). The service A makes a call to an external api service (called B). In my tests, i would like to mock all the calls to the service B as it will not be availabed when i build my app on jenkins for example.
I tried examples from the doc by it doesn't work : 
`Given that calling "https://service-B" will return a status OK_200 and:`
    `"""`
   `some data here`
    `"""`
    `When a user send on "/service-A":`
     ` """`
     ` method: POST`
      `body:`
       ` payload:`
          `some data here`
      `"""`
    `Then it receives a status CREATED_201`

This doesn't work me at first because the service B is called by my service A and not directly in the tests as this example : https://github.com/Decathlon/tzatziki/blob/6e0b97d0323cbd8fdd370d3e9d4773172842568b/tzatziki-http/src/test/resources/com/decathlon/tzatziki/steps/http.feature#L3

To resolve the issue, in my steps initializer class , i added this system env : 
`System.setProperty("ENDPOINT", MockFaster.url());`
this variable is used by my code to make the call to B (in application-test.yaml as i'm using spring boot) : 
`host-service-B: ${ENDPOINT}/service-B-resource`
and in the gherkin, i'm not using urls prefixed with http and https. So in the last test we have to change just the first line : 
`Given that calling "/service-B" will return a status OK_200 and:`
Using this, all the api calls will be intercepted by the MockFaster.url() and mocked.

I don't know if this is the good way to use the library ?
If so, i think that this part have to be added to the documentation ?

**Comment by brian-mulier** [comment]
Hello !

What you said is correct. In fact most of the times you will want to have an environment-dependant endpoint which is why this is not that much of an issue. What you can do is consider MockFaster.url() as an API Portal. If you are not using such portal, you can simply specify the whole endpoint as an environment variable which will allow you to : 

- on production side, always specify the whole endpoint (including http(s)) into the environment variable
- on test side, specify MockFaster.url() + a relative endpoint into the environment variable

When it comes to understanding why it is working like that, MockFaster is keeping a memory of host-based relative paths. To match against those path, the only way is to make an HTTP call against MockFaster, else you will end up calling the real internet :)

The steps to call an endpoint are mapping an host-based path to a MockFaster URL if it is matching a mocked path, else it is reaching out Internet as shown below :
![image](https://user-images.githubusercontent.com/69139631/151549995-e8475556-8777-4d96-b2ea-b8a71f4fcc45.png)
![image](https://user-images.githubusercontent.com/69139631/151550158-96b683ac-c1f2-43b7-8613-b150431ef7ca.png)

It is effectively unclear in the documentation and should be added, although technically an understandable mechanism.
Feel free to add the documentation through a PR if you want. Else I'll do it later.

**Reply by mamdouni** [reply]
thanks for the explanation @brian-mulier 🥇 it helps me to understand how the library works internally
i'll take in charge the PR

**Comment by mimfgg** [comment]
super explanation :)

But yes, it is true that the urls starting with http:// will be automatically remapped to their local mocks starting with MockFaster.url() so it could be a bit ambiguous as a feature ... maybe we could remove this to make it more intuitive ...

It is currently not really possible to mock a random url since we don't control the http client used by your application. This would be possible though ... we could encapsulate the RestTemplate or WebFlux builders in the spring module ... maybe a nice feature to implement!

**Reply by mamdouni** [reply]
yes, i agree with you. but it will be difficult as there is a lot of http clients. In my case, we are using the third part library that uses apache http client :exclamation:

**Reply by mimfgg** [reply]
I have a PR for the RestTemplate and WebClient https://github.com/Decathlon/tzatziki/pull/19 but it should be easy to apply the same logic on your custom client.

**Comment by mamdouni** [comment]
Thank you for the response, i have another question in the same context and i don't know if it needs a new discussion ?
Actually, when calling internet (without using mocks) i get a timeout exception. Here is my test case : 
![image](https://user-images.githubusercontent.com/61866853/151667595-bd36a5c2-6dad-457f-8abc-2e69c4738dc1.png)

Doing the same test on a new project without the spring module works perfectly : 

![image](https://user-images.githubusercontent.com/61866853/151667662-2cc83f0b-7438-4a13-a0bc-7e0fb6ba4ef2.png)

The first test fails (timeout) after calling the RestAssured.request method : 
![image](https://user-images.githubusercontent.com/61866853/151667751-a21e0b53-e0d2-4727-bbbe-97fac341d699.png)
This method calls the host using the RestAssured.port which is -1 by default -> reach out the internet
This configuration will be overriden after lauching the test by **Before** method in **SpringSteps** : 
![image](https://user-images.githubusercontent.com/61866853/151667918-ba469382-12d2-4f2f-a006-7eda5efe7c2a.png)

So when calling google.com, the request will be redirected to google.com:{Spring_Random_Port} and not to  google.com:80.
Is this a bug in the library ? or a manner to say, during unit tests we cannot do external calls ?

**Reply by brian-mulier** [reply]
I may be wrong since I can't check the code until Monday but I think the RestAssured port is used only in case of specifying a relative path in order to be able to reach out your app's controllers. This way you can call and test your own API. I can't get why the test is wrong there. Is there any dependency difference around Tzatziki between your fresh project and the old one ?

I will try to reproduce it on Monday if you aren't able to figure out what's wrong. @mimfgg may have an idea about what's going on ?

**Reply by brian-mulier** [reply]
And just to make it clear : you can reach out internet through Tzatziki. I'm actually doing it for my integration tests on a pre-production environment.

**Reply by mimfgg** [reply]
that might be a bug ... because reaching out to the internet should work, like @brian-mulier said, we do it. I will find why it doesn't behave like it should and add that to the other PR :)

**Reply by mimfgg** [reply]
you were right, the RestAssured port was interferring with the network call. I have fixed it in the PR!
Thanks for debugging and reporting it!

**Comment by mimfgg** [comment]
I have released the version 1.0.5 fixing all the problems raised in this thread!

**Reply by mamdouni** [reply]
good news,
Thank you guys for your reactivity and for fixing the bug. I'll test the new version.

**Reply by mimfgg** [reply]
try the version 1.0.8 then :)

**Reply by mamdouni** [reply]
yep, i tested it with 1.0.8 and it works like a charm :ok_hand:
![image](https://user-images.githubusercontent.com/61866853/152017986-7922705c-2dcb-4e33-befd-70fde7e97d3c.png)
